### PR TITLE
make nulls direction configuable for FullSortingMergeJoin

### PR DESCRIPTION
**Pull request by loudongfeng**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add a flag for SMJ to treat null as biggest/smallest. So the behavior can be compitable with other SQL systems, like Apache Spark.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Review by vdimir — COMMENTED** [review]
Thanks for your work on ClickHouse!

Could you please add more details about use-case?
Generally, `NULL`s should not affect join, because they are never joined.
In case you would like to have specific `NULL`s direction _after_ join, you may just add `ORDER BY` with specific modifier, but join itself does not guarantee an order

**Comment by loudongfeng** [comment]
Thanks for you attention.
 Actually, the use case is not for ClickHouse user, but for developers who want to use ClickHouse as a vectorization library.
For example, in Apache Gluten, we accelerate spark's Sort Merge Join by offloading it into ClickHouse 's SortingStep and JoinStep with FullSortingMergeJoin. The SortingStep offloading is optional(which means sorting can be done in spark's SortExec in nulls smallest order and then pass data to CH's FullSortingMergeJoin), so we need assign null orders for FullSortingMergeJoin in that case.
  From ClickHouse's point of view, it's just an internal API richness.

File: src/Processors/Transforms/MergeJoinTransform.cpp [hunk 1]
```diff
@@ -356,7 +359,7 @@ void MergeJoinAlgorithm::consume(Input & input, size_t source_num)
     cursors[source_num]->setChunk(std::move(input.chunk));
 }
 
-template <JoinKind kind>
+template <JoinKind kind, NullOrder nullOrder>
```
Comment (vdimir) [inline src/Processors/Transforms/MergeJoinTransform.cpp:362]: Can't we pass direction as a value?

File: src/Core/Settings.h [hunk 2]
```diff
@@ -880,6 +880,7 @@ class IColumn;
     M(Int64, ignore_cold_parts_seconds, 0, "Only available in ClickHouse Cloud", 0) \
     M(Int64, prefer_warmed_unmerged_parts_seconds, 0, "Only available in ClickHouse Cloud", 0) \
     M(Bool, iceberg_engine_ignore_schema_evolution, false, "Ignore schema evolution in Iceberg table engine and read all data using latest schema saved on table creation. Note that it can lead to incorrect result", 0) \
+    M(Bool, nulls_biggest_in_smj, true, "Treat nulls as biggest in sort. Used in sort merge join for compare null keys.", 0) \
```
Comment (vdimir) [inline src/Core/Settings.h:883]: Let's not expose it to user

**Review by vdimir — COMMENTED** [review]
Let's then simply add an argument to the MergeJoinAlgorithm and define it as a constant somewhere outside. This will meet your requirements without introducing significant changes. Especially, adding it as a template argument is questionable since it increases the binary size, but `nullableCompareAt` already has a runtime argument now.

File: src/Processors/Transforms/MergeJoinTransform.cpp (hunk 1 above)
Comment (loudongfeng) [reply src/Processors/Transforms/MergeJoinTransform.cpp]: fixed

File: src/Core/Settings.h (hunk 2 above)
Comment (loudongfeng) [reply src/Core/Settings.h]: fixed

**Comment by loudongfeng** [comment]
Hi vdimir, thanks for your review.
I have remove the template argument as you suggested.
But i'm not sure how to fix this one:
```
Let's then simply add an argument to the MergeJoinAlgorithm and define it as a constant somewhere outside.
```
Do you mean no changes in JoinStep,QueryPipelineBuilder,MergeJoinTransform etc and only add a runtime argument for MergeJoinAlgorithm?
But how to use a constant to set null_direction_hint, since it is unchangable.
Sample code that gluten uses to offloading SMJ to ClickHouse library:
```c++
JoinPtr smj_join = std::make_shared<FullSortingMergeJoin>(table_join, right->getCurrentDataStream().header.cloneEmpty());
QueryPlanStepPtr join_step
                 = std::make_unique<DB::JoinStep>(left->getCurrentDataStream(), right->getCurrentDataStream(), smj_join, 8192, 1, false);
```

**Comment by vdimir** [comment]
Yes I mean just like that

```cpp
auto null_direction = getNullDirection();

JoinPtr smj_join = std::make_shared<FullSortingMergeJoin>(table_join, right->getCurrentDataStream().header.cloneEmpty(), null_direction);
QueryPlanStepPtr join_step
                 = std::make_unique<DB::JoinStep>(left->getCurrentDataStream(), right->getCurrentDataStream(), smj_join, 8192, 1, false);
```


Also, to be honest I don't fully understand how it can be useful, because handling nulls is `FullSortingMergeJoin` implementation detail, and if you need particular order after that you may want add specific step further.

**Comment by loudongfeng** [comment]
Particular null order after join does not matter, but null order before Join matters.
I find nulls smaller/bigger matters after digging a result mismatch issue in gluten. Here is the case
```sql
-- a simple table
CREATE TABLE item (
  i_current_price DECIMAL(7,2),
  i_category STRING)
USING parquet;

-- data with nulls
insert into item values
(null,null),
(null,null),
(0.63,null),
(0.74,null),
(null,null),
(90.72,'Books'),
(99.89,'Books'),
(99.41,'Books');


-- the following SQL will use SMJ with join type Left Outer
-- expected result is 2, but gluten returns 0 when offloading to ClickHouse
-- notice there are nulls in join key 'i_category'
SELECT  count(*) cnt
FROM item i
where
i.i_current_price > 1.0 *
  (SELECT avg(j.i_current_price)
  FROM item j
  WHERE j.i_category = i.i_category
 )
```

The reason is sort node's order is asc and nulls first (both in Spark and ClickHouse's SortStep), but FullSortingMergeJoin  joins streams treating data in asc order and nulls last.  FullSortingMergeJoin will stop when find left/right table start with null, I think. Simply changing null direction in  FullSortingMergeJoin can fix it.

**Comment by loudongfeng** [comment]
CI failures seems unrelated. Ready for review. 
Thanks, @vdimir

**Review by vdimir — APPROVED** [review]

**Comment by vdimir** [comment]
Thanks for your effort on the PR!

**Pull request merged**